### ISSUE TITLE
Can we allow some statements at module level?

### ISSUE THREAD

**Issue by jclark** [ballerina-platform/ballerina-spec#660]
Can/should we allow some (or all) kinds of statements at module level with the semantic that it is equivalent to putting the statements in an init function?

Then our hello world would become:

```
import ballerina/io;
io:println("Hello World!");
```

**Comment by jclark** [comment]
At the moment the principle is that, with the exception of import statements, you can put module-level constructs in any order they want. You do not have to declare a top-level function or variable before you reference it. Initialization are reordered as necessary so that no variable is used without first being initialized.

I don't think we should try to combine this ordering model with the very different ordering model needed for multiple statements without very careful consideration. At the moment a file has two sections: imports and then other module-level declarations. One possibility would be to say a file has three sections:

1. imports
2. module-level declarations - order insignificant
3. statements - order significant - only some kinds of statement would make sense; for example, you could not have a local variable declaration statement, because that would be ambiguous with a module-level declaration

The probably only one file in a module should have a statements section.

There is a connection with #35.

I think this is too much to try to do for the initial release of Swan Lake.

**Comment by sanjiva** [comment]
I don't think we should have structured sections like that. Languages that allow this stuff don't typically do that IIRC .. and to have a REPL environment and a notebook environment supported, you have to allow that.

Can we remove the imports-must-come-first-before-all decls requirement? The primary need is for imports to come before its used, beyond that why do we care? For code styling its best to put them at the top for sure, but is it really beneficial to require that at a language level?

If we do that we can allow imports, module level decls and statements to come in any order. Such statements in each file are lexically ordered and multiple files in a module can be ordered (as discussed already). Then all of that code goes into the module init and we have a well-defined execution order.

Local variables would indeed not be possible- if one wants that then the code would have to be moved to the init function. There is also a question of whether init() runs first or the statements run first. It seems init() should go and then the statements.

**Comment by jclark** [comment]
I made a separate issue #662 for relaxing grammar of imports.

My concern is not whether we can define an execution order (we certainly can); it is that we would have two inconsistent treatments of ordering of syntactic constructs that occur at module-level.

**Comment by jclark** [comment]
We discussed this today. Conclusion is that the real problem is to define how Ballerina  can be used in a shell-like environment. Examples:

- a shell - statements or declarations get executed incrementally as they are input
- shell script - `#!/usr/bin/ballerina`  should similarly to the shell
- notebook - e.g. Jupyter
- REPL - here we also have the print aspect and we probably want to handle expressions

It is generally desirable to have a looser approach for this sort of situation than we have for modules of a program. For example, in a shell you should be able to redefine a function, but a module certainly should not allow two instances of the same function name.

We can potentially unify this with single-file Ballerina programs.

We probably need a different file extension, rather than relying on whether the file is part of a project, since projects may want to include Ballerina scripts.
